feat: grouping & ungrouping shapes

pengx17 opened the pull request:
https://user-images.githubusercontent.com/584378/207405558-741220c7-3ceb-46c4-94b0-4e12b4f341f4.mp4

Add new context menu actions "group" & "ungroup".
- Ungroup will only show if any of the selected shape is in a group or is a group.
- group will show if selected more than one shape

tiensonqin commented:
@pengx17 Such a great PR! 👍 🎉 

I QAed both grouping and ungrouping shapes, they're working great. It'll be nice if there's a way to allow users to enter and exit the grouping mode, this is a screenshot from whimsical:
![CleanShot 2022-12-14 at 11 29 55](https://user-images.githubusercontent.com/479169/207498825-59bbaf32-b32e-45f9-987d-c792bfed5aaa.png)

I love the Enter group and ungroup actions when the group has been selected in whimsical, they also make it very easy to use by typing Enter to enter the group (then users can edit the inner shapes) and ESC to restore the group.

I also noticed a crash bug when removing an arrow.
https://www.loom.com/share/43f74757a1aa4c81a90585477f5e528c

pengx17 commented:
@tiensonqin  Crash issue should be fixed now.

pengx17 commented:
> It'll be nice if there's a way to allow users to enter and exit the grouping mode, this is a screenshot from whimsical

I will think about how to implement this behavior. We can do this in another PR right?

tiensonqin commented:
@pengx17 Sure, we can improve the UX in another PR.

tiensonqin commented:
Confirm that the crash bug has been fixed! 👍 

Another thing we can improve is to allow selecting a group by clicking the empty area (not shapes) inside the group selection.
https://www.loom.com/share/283d919e6b36479790198393107ec37b

tiensonqin reviewed (approved):
LGTM

pengx17 commented:
> Another thing we can improve is to allow selecting a group by clicking the empty area (not shapes) inside the group selection.

I removed this function because I noticed that clicking the empty area will not select the group in (beta.)tldraw.com. I have added back because I do also think it is more intuitive.

sprocketc reviewed (approved):
Looks good and works great!

Some suggestions for this (or for follow up PRs)
> typing Enter to enter the group (then users can edit the inner shapes) and ESC to restore the group. mod+g to group and mod+shift+g to ungroup.

I agree with @tiensonqin. Ctrl+G and Ctrl+Shift+G for grouping/ungroupimg is common. Double clicking to enter a group would also be nice.

The group bounds should either overlap with the actual bounds of the included shapes, or the inner bounds should be hidden in my opinion. Let's wait for @scheinriese's input on this.

cannibalox commented:
> > Another thing we can improve is to allow selecting a group by clicking the empty area (not shapes) inside the group selection.
> 
> I removed this function because I noticed that clicking the empty area will not select the group in (beta.)tldraw.com. I have added back because I do also think it is more intuitive.

I believe he prevents the `select when clicking in empty area` to avoid issues when group boundaries also overlap/encompass separate elements or other groups ?
![Maxthon_tldraw_-_Maxthon_20221215_1877](https://user-images.githubusercontent.com/4605693/207845813-23315bbf-d762-4aa1-a78a-d1d61bdceef8.png)

pengx17 commented:
> I believe he prevents the select when clicking in empty area to avoid issues when group boundaries also overlap/encompass separate elements or other groups ?

This is really a very good point! What do you think @tiensonqin ?

tiensonqin commented:
@cannibalox good point!
@pengx17 Maybe we can always select the top front shape or group?

pengx17 commented:
> @pengx17 Maybe we can always select the top front shape or group?

It should already be the case.

The ordering of a group vs a normal shape can be changed by the user. A shape could be behind a group but still visually not being covered by a group. In this case, the user maybe surprised that the shape cannot be selected.

tiensonqin commented:
> It should already be the case.
Can confirm this 👍 

> A shape could be behind a group but still visually not being covered by a group. In this case, the user maybe surprised that the shape cannot be selected.

Indeed, the good news is that the user can still move or ungroup the group to see the shape behind it.
